### ISSUE TITLE
new_class includes things that don't end in a dot

### ISSUE THREAD

**Issue by sogaiu** [oakmac/tree-sitter-clojure#27]
The current rule for new_class is:
```
new_class: $ => /([a-zA-Z_]\w*\.)(\w+\.)*/,
```
This will also pick things up that don't end with a dot like: `java.util.BitSet`

Assuming we only want things here that end with a dot, may be the following works better?
```
new_class: $ => /([a-zA-Z_]\w*)(\.\w+)*\./,
```

**Comment by sogaiu** [comment]
Parsing `java.util.BitSet` using the suggested alternative regular expression leads to:
```
(program [0, 0] - [1, 0]
  (interop [0, 0] - [0, 10]
    (new_class [0, 0] - [0, 10]))
  (symbol [0, 10] - [0, 16]))
```
This may be because `symbol_chars` is not defined as allowing dots:
```
_symbol_chars: $ => /[a-zA-Z\*\+\!\-_\?][a-zA-Z0-9\*\+\!\-_\?\':]*/,
```
The [reference](https://clojure.org/reference/reader#_symbols) quoted in grammar.js and a local repl suggest the dot should be included:
> '.' has special meaning - it can be used one or more times in the middle of a symbol to designate a fully-qualified class name, e.g. java.util.BitSet, or in namespace names. Symbols beginning or ending with '.' are reserved by Clojure. Symbols containing / or . are said to be 'qualified'.
```
user=> (symbol? 'java.uti.BitSet)
true
```
Modifying `_symbol_chars` to be:
```
_symbol_chars: $ => /[a-zA-Z\*\+\!\-_\?][a-zA-Z0-9\*\+\!\-_\?\':\.]*/,
```
yields the following result when parsing:
```
(program [0, 0] - [1, 0]
  (symbol [0, 0] - [0, 16]))
```
Note that the modified regular expression would allow things that end in a dot to work as symbols, which doesn't seem correct according to the reference (though the local repl permits it).

I think parsing of things of the form: `(java.util.Date.)` breaks with the suggested modification, so perhaps it needs more work.

**Comment by sogaiu** [comment]
Note that the reference also states:
> Symbols beginning or ending with ':' are reserved by Clojure. A symbol can contain one or more non-repeating ':'s.

IIUC, the current regular expression for `_symbol_chars` allows things that end in a colon, so may be that should be changed appropriately as well?

**Comment by sogaiu** [comment]
Another case that seems related.

Parsing `clojure.core/inc` using the current grammar yields:
```
(program [0, 0] - [1, 0]
  (interop [0, 0] - [0, 8]
    (new_class [0, 0] - [0, 8]))
  (symbol [0, 8] - [0, 16]
    (qualified_symbol [0, 8] - [0, 16])))
```
This doesn't seem correct either.

**Comment by sogaiu** [comment]
I've been trying out the following modified version:

```
_symbol_chars: $ => /[a-zA-Z\*\+\!\-_\?&]([a-zA-Z0-9\*\+\!\-_\?&\':.]*[a-zA-Z0-9\*\+\!\-_\?&\'])?/,
```

Locally, this (possibly with modifications elsewhere) appears to handle:

* `clojure.core/inc`
* `java.util.BitSet`
* `&form`

**Comment by kolja** [comment]
I submitted a pull request that can handle these cases: https://github.com/oakmac/tree-sitter-clojure/pull/30
Another thought: Apparently "$" is also valid in class names.

**Comment by sogaiu** [comment]
Yes, I think that is correct, '$' is used for specifying inner classes, so I think it would be more comprehensive to account for that.

----

BTW, thanks for asking the question on StackOverflow -- investigating it really helped to improve my understanding of how tree-sitter works (though I am sure I still have much to learn).

In trying to get this repository's version of the grammar to work with the corpus from Tavistock's as well as tonsky's test data, I struggled a lot and decided to try to write a less ambitious grammar (at least initially).  This reduced grammar appears to handle clojure and clojurescript's source code pretty well (no outright occurences of ERROR, for example).  It doesn't do things like recognize function definitions or interop as things separate from lists, however.

I briefly compared the parsing performance of the grammar in this repository with that in Tavistock's (as well as the reduced one) and noticed that this one was about 3 times slower.  So depending on the use case, it might make sense to have different levels of "featurefulness" in different grammars (e.g. I used an even more reduced one to implement rainbow parens in vscode).

The experimentation with the various grammars and attempting to answer the SO question has helped to make some other factors clear.  For example, control of tokenization is quite important to get appropriate things on to the stack used for building an appropriate tree.  It has happened repeatedly that what I want to get tokenized one way ends up tokenized another way and initially it appeared that there was no good way to influence this process.

Combing the documentation and trial and error revealed that reducing the use of literal regular expressions and instead using literal strings in combination with `seq`, `choice`, etc. plus `prec` appear to give one some ability to affect which tokenization rule is actually used.  (In case it's not clear already, I found that the `--debug-graph` option to the tree-sitter `parse` subcommand can be helpful in seeing the details -- the results end up in a file named `log.html` which appears to have output similar to what one may see in some of the tree-sitter talks.)

It wasn't obvious to me that one can place `prec` within a call to `token`, but it appears to help in some cases.  Also, scanning other grammars, I see that it is used in a small number of cases.

If you'd like to take a look at the reduced grammar, it is at:

https://github.com/sogaiu/tree-sitter-clojure

I also created a #tree-sitter channel on slack.

**Comment by kolja** [comment]
Your answer on StackOverflow was an eye-opener, thank you again! I have those changes applied to my fork of this project and it's working well for me (parsing the response from an unrepl server).
I checked out your reduced grammar, too. Very nice. I wonder why the tree-sitter project doesn't list a repository with a Clojure grammar. This could be a way to bundle everyones efforts...

**Comment by sogaiu** [comment]
Regarding the tree-sitter project not yet mentioning a Clojure grammar, perhaps you are referring to the section titled "Parsers for these languages are in development" at the ["top page"](http://tree-sitter.github.io/tree-sitter/) for tree-sitter?

**Comment by sogaiu** [comment]
@kolja Is what you are working on [repl-aliance](https://github.com/kolja/repl-alliance)?

On a side note, I've also used borkdude's [edamame](https://github.com/borkdude/edamame) to achieve some of the same ends I've been trying to use tree-sitter for.  Have you had a look at that?

**Comment by kolja** [comment]
@sogaiu yes, I am parsing Clojure (EDN?) data returned from prepl and unrepl servers. Edamame is interesting, but repl-alliance is a neovim plugin and it's all written in Lua: Version 0.5 of neovim (which is in beta now) will have support for Tree-sitter. 
It is entirely possible though that I am missing some important point and doing things the wrong way. It's a learning experience for me.
That said, with a few changes to this grammar over on my fork the server responses parse without errors.

**Comment by sogaiu** [comment]
Thanks for the explanation.

I'm coming to see that tree-sitter stuff being available via C / wasm opens up more tooling opportunities.

----

On a side note, is repl-alliance something I should be able to use now with the socket repl or prepl?

I tried a bit but am getting `ECONNREFUSED` after starting a socket repl via a terminal, then filling in some values in nvim with:
```
:RAconnect
```
and then:
```
:RAeval (+ 1 1)
```
yields:
```
Error executing luv callback:
/home/user/.vim/plugged/repl-alliance/lua/repl.lua:158: ECONNREFUSED
```

**Comment by kolja** [comment]
I appreciate your trying repl-alliance. You are likely the first person to use it! It's a work in progress, but what you did is absolutely correct. There is a bug in `:RAconnect`. I fixed it just now.
Could you please run `:PlugUpdate` and try again?
I think you were getting `ECONNREFUSED` because `:RAconnect` didn't pick up the port that you entered and tried to connect on the default port, which is 7322.

**Comment by sogaiu** [comment]
Thanks for the update!

It worked :+1: 

```
:RAeval (+ 1 1)
```

yielded:

```
user=> (+ 1 1)
user=> 
2

user=> 
```

**Comment by kolja** [comment]
Very good! The regular socket-repl output looks weird. It's better with nrepl and prepl.
